### PR TITLE
[RFC] Improve prompt part of denite.

### PR DESCRIPTION
**Pull request by lambdalisue**
Re-opend #40

[Migration note](https://github.com/Shougo/denite.nvim/wiki/Migration-note)

With this PR, you can

- Define custom mappings with multiple keys like `<C-a><C-b>`
- Define `expr` like mappings to execute Vim script within a mapping
- Define `noremap` like mappings to prevent infinit recursive mappings
- Define `nowait` like mappings
- Move cursor left and right with `<Left>/<Right>`, insert characters at the cursor position
- Change Insert/Replace mode with `<Insert>`
- Paste text from a specified register with `<C-r>`
- Paste text from a default register (Not mapped)
- Yank text to a specified register (Not mapped)
- Yank text to a default register (Not mapped)
- Enter digraph (e.g.  å, あ) by `<C-K>` (See `:help digraph`)
- Enter non-digit literal (e.g. ^V, ^M) by `<C-Q>` (See `:help c_CTRL-Q`)
- Use command-line history with `<Up>/<Down>/<C-p>/<C-n>`
- Use `timeout` and `timeoutlen` to regulate the behaviour

The prepared actions are listed at

- https://github.com/lambdalisue/denite.nvim/blob/neovim-prompt/rplugin/python3/denite/prompt/action.py#L319
- https://github.com/lambdalisue/denite.nvim/blob/neovim-prompt/rplugin/python3/denite/ui/action.py#L59

And default mappings are listed at

- https://github.com/lambdalisue/denite.nvim/blob/neovim-prompt/rplugin/python3/denite/ui/action.py#L81

So denite would be more flexible and powerful with neovim-prompt :+1:

### The expr mapping

Users can define `expr` like mappings. The following example map `;` to use [pinkyless](https://github.com/lambdalisue/pinkyless.vim)'s stickyshift feature.

```vim
call denite#custom#map('insert', ';', 'pinkyless#stickyshift#enter(";")', 'noremap expr')
```

Then `;a` become `A` in insert mode.

### Backward incompatibilities

- Mapping name has changed like from `suspend` to `<denite:suspend>`.
- `g` in normal mode become `gg`
- `<C-r>` became more similar to the Vim's builtin. User need to assign `<denite:paste_from_default_register>` if he/she like the previous behaviour
- Existing text is not shown in `<C-j>` while it is not possible to set an initial caret position in `input()`.
- Text will be inserted at the caret position rather than the end of the text after `<C-j>` 

If you find any difference other than above, please let me know.

### ToDo
- [x] Add neovim-prompt as git subtree
- [x] Create denite UI which use neovim-prompt
- [x] Support actions
- [x] Support async feature
- [x] Support resume feature
- [x] Put `<C-j>` back (with a bit different behaviour)
- [x] Rename `<prompt:xxxxx>` to `<denite:xxxxx>`
- [x] Close denite whenever user press `<C-c>`
- [x] Write documents
- [x] Refactoring/Lint fix
- [x] Remove `<Tab>` and `<S-Tab>` mappings (https://github.com/Shougo/denite.nvim/commit/1bc1bf47dd3f22728aec254f89d3889b0b27182b)
- [x] Write migration note

**This PR may break the current denite while it touched and changed a core part of denite so please help us with testing this branch**

### CODE REVIEW

**Comment by Shougo** [comment]
Thanks.  I will check it later.

**Comment by rafi** [comment]
@lambdalisue this is awesome! 🎆 
Found an issue: after doing 'preview', Denite window loses neovim-prompt. There should be support for actions that _"do not hide the Denite window"_, probably a feature that will be added to _kind actions_, but currently only 'preview' uses this behavior.

**Comment by lambdalisue** [comment]
Additional todo

* Put back C-j
* Remove preview from default insert mapping
* Use denite prefix rather than prompt for future
* Some refactoring

**Comment by lambdalisue** [comment]
@rafi So I don' need to fix the behavior in this pr rihgt?

**Comment by Shougo** [comment]
> Found an issue: after doing 'preview', Denite window loses neovim-prompt. There should be support for actions that "do not hide the Denite window", probably a feature that will be added to kind actions, but currently only 'preview' uses this behavior.

> Edit: Seems to happen in master branch as-well..

Please create new issue.  I will check it.

**Comment by rafi** [comment]
@lambdalisue seems there's conflicts. can you rebase or merge `master`?

**Comment by lambdalisue** [comment]
@rafi I'm waiting #86 to be merged while that would conflict.

**Comment by lambdalisue** [comment]
I changed mind and merge the current master into this PR while this PR solve a same issue of the #86.

File: doc/denite.txt [hunk 1]
```diff
@@ -1025,14 +1269,6 @@ A: >
...
 	:Denite -resume -select=-1 -immediately
 
-Q: I want to move to normal mode by "jj".
-
-A: You can define the mode for it.
->
-	call denite#custom#map('insert', 'j', 'enter_mode:j')
-	call denite#custom#map('j', 'j', 'enter_mode:normal')
-	call denite#custom#map('j', '<Space>', 'insert_word:j')
->
```
Comment (lambdalisue) [inline doc/denite.txt]: I removed this section while the following works with this PR

```
call denite#custom#map('insert', 'jj', '<denite:enter_mode:normal>', 'noremap')
```

Comment (Shougo) [inline doc/denite.txt]: OK.  Thanks.

**Comment by Shougo** [comment]
Hm...  Conflict files...

**Comment by lambdalisue** [comment]
Probably, this PR is ready to merge :tada:. Please give me :+1: if you don't see any issues on this branch.

**Comment by pocari** [comment]
@lambdalisue  This is a good feature!

But, an error occurred in this way.

1. :Denite grep
2. Search by any phrase.
3. close denite window(by `<ESC>`)
4. :Denite grep -buffer-name=aaa

```
function denite#helper#call_denite[14]..denite#start[15].._denite_start の処理中にエラーが検出されました:
行    1:
[denite] Traceback (most recent call last):
[denite]   File "/Users/pocari/.cache/dein2/.cache/test_vimrc/.dein/rplugin/python3/denite/__init__.py", line 32, in start
[denite]     self.__uis[buffer_name] = Default(self.__vim)
[denite]   File "/Users/pocari/.cache/dein2/.cache/test_vimrc/.dein/rplugin/python3/denite/ui/default.py", line 53, in __init__
[denite]     weakref.proxy(self)
[denite]   File "/Users/pocari/.cache/dein2/.cache/test_vimrc/.dein/rplugin/python3/denite/ui/prompt.py", line 17, in __init__
[denite]     self.action.unregister('prompt:accept')
[denite]   File "/Users/pocari/.cache/dein2/.cache/test_vimrc/.dein/rplugin/python3/denite/prompt/action.py", line 75, in unregister
[denite]     raise e
[denite]   File "/Users/pocari/.cache/dein2/.cache/test_vimrc/.dein/rplugin/python3/denite/prompt/action.py", line 72, in unregister
[denite]     del self.registry[name]
[denite] KeyError: 'prompt:accept'
[denite] Please execute :messages command.
```

**Comment by Shougo** [comment]
Pending until the bug is fixed.

**Comment by lambdalisue** [comment]
@pocari Thanks. I'll investigate.

**Comment by pocari** [comment]
@lambdalisue Errors no longer occured 👍

**Comment by lambdalisue** [comment]
https://github.com/Shougo/denite.nvim/wiki/Migration-note

**Comment by AlexMasterov** [comment]
**Neovim:** works fine.
**Vim 8**:
```viml
call dein#add('lambdalisue/denite.nvim', {'rev': 'neovim-prompt', 'on_cmd': 'Denite'})
```
:Denite
```viml
Error detected while processing <...>\dein\repos\github.com\lambdalisue\denite.nvim\autoload\denite\init.vim:
line   56:
E127: Cannot redefine function denite#init#_initialize: It is in use
Error detected while processing function denite#helper#call_denite[14]..denite#start[11]..denite#initialize[1]..denite#init#_initialize[10]..denite#vim#_initialize:
line   21:
E117: Unknown function: denite#init#_variables
```

**Comment by Shougo** [comment]
Pending until the error is fixed in Vim8.

**Comment by lambdalisue** [comment]
Done 👍

**Comment by Shougo** [comment]
Merged!

**Comment by wsdjeg** [comment]
@lambdalisue thanks for this PR, I also do something like it in my plugin [vim-chat](https://github.com/wsdjeg/vim-chat/blob/master/autoload/chat/qq.vim#L326-L445) some time ago, but I do it in viml instead of python. BTW I found a nice feature of your PR, in vim8 the cursor at the end of cmdline has been hidden. how do you do?

**Comment by lambdalisue** [comment]
> in vim8 the cursor at the end of cmdline has been hidden. how do you do?

Nothing. I actually gave up about that in neovim (and I didn't notice that Vim 8 hide the cursor) so I've no idea how can I hide the cursor at the end of the prompt (If you find a way, let me know).

Additionally, I remember that I had same problem in Vim so maybe recent Vim changed some feature?

**Comment by wsdjeg** [comment]
it works well for me in vim8, and the cursor is hidden, here is two pics, first one is in neovim.

**Comment by lambdalisue** [comment]
Yeah I noticed that but I don't know why. I haven't done anything special for that.

**Pull request merged**